src/main/java/com/google/step/snippet/servlets/SearchServlet.java

```diff
@@ -79,13 +84,53 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
     }
     String query = encodeValue(param);
     List<Card> allCards = new ArrayList<>();
-    for (Client client : clients) {
-      String link = getLink(client.getCseId(), query);
-      Card card = client.search(link);
-      if (card != null) {
-        allCards.add(card);
-      }
+    ExecutorService executor = Executors.newFixedThreadPool(3);
+    List<Callable<Card>> cardCallbacks =
+        Arrays.asList(
+            () -> {
+              String w3Link = getLink(w3Client.getCseId(), query);
+              if (w3Link != null) {
+                return w3Client.search(w3Link);
+              }
+              return null;
+            },
+            () -> {
+              String stackLink = getLink(stackClient.getCseId(), query);
+              if (stackLink != null) {
+                return stackClient.search(stackLink);
+              }
+              return null;
+            },
+            () -> {
+              String geeksLink = getLink(geeksClient.getCseId(), query);
+              if (geeksLink != null) {
+                return geeksClient.search(geeksLink);
+              }
+              return null;
+            });
+    /* thread process completion takes roughly 1.5 to 2.5 seconds, thus a timeout of
+     * 3 seconds was chosen.
+     */
+    try {
+      executor.invokeAll(cardCallbacks, 3L, TimeUnit.SECONDS).stream()
+          .map(
+              future -> {
+                try {
+                  return future.get();
+                } catch (CancellationException | ExecutionException | InterruptedException e) {
+                  return null;
+                }
+              })
+          .forEach(
+              (c) -> {
+                if (c != null) {
+                  allCards.add(c);
+                }
+              });
+    } catch (InterruptedException | NullPointerException | RejectedExecutionException e) {
+      return;
     }
```

Comment:
I don't think it should be returning here? If an error is caught, SearchServlet should still redirect to the SRP page just with no cards.